Implement Observable class and subscribe method.

keithamus opened the pull request:
#### dbafcabcd37eb832eddbc347358f355fc7357c5f
<pre>
Implement Observable class and subscribe method.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276572">https://bugs.webkit.org/show_bug.cgi?id=276572</a>

Reviewed by Ryan Reno.

This implements the base `Observable` class, as well as the `subscribe`
method. In addition this adds the `Subscriber` internal class allows
Observables to be subscribed to, and some additional machinery such as
`SubscribeOptions`, `SubscriptionObserver` and so on.

This also includes a fix to IDL generation to correctly generate callback
function + dictionary unions (<a href="https://github.com/whatwg/webidl/pull/1353).">https://github.com/whatwg/webidl/pull/1353).</a>

This does not implement any of the prototype methods on Observable,
(other than subscribe) such as `takeUntil`, `filter`, `finally` and so
on.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-takeUntil.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-takeUntil.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-takeUntil.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any.worker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/JSDOMConvertUnion.h:
* Source/WebCore/bindings/js/JSSubscriberCustom.cpp: Added.
(WebCore::JSSubscriber::visitAdditionalChildren):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(AreTypesDistinguishableForOverloadResolution):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackWithFunctionOrDict.cpp: Added.
(WebCore::JSTestCallbackWithFunctionOrDict::JSTestCallbackWithFunctionOrDict):
(WebCore::JSTestCallbackWithFunctionOrDict::~JSTestCallbackWithFunctionOrDict):
(WebCore::JSTestCallbackWithFunctionOrDict::handleEvent):
(WebCore::JSTestCallbackWithFunctionOrDict::visitJSFunction):
(WebCore::toJS):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackWithFunctionOrDict.h: Added.
(WebCore::toJS):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestCallbackOrDict.idl: Added.
* Source/WebCore/dom/Observable.cpp: Added.
(WebCore::Observable::create):
(WebCore::Observable::subscribe):
(WebCore::Observable::makeSubscriber):
(WebCore::Observable::Observable):
* Source/WebCore/dom/Observable.h: Added.
* Source/WebCore/dom/Observable.idl: Added.
* Source/WebCore/dom/SubscribeOptions.h: Added.
* Source/WebCore/dom/SubscribeOptions.idl: Added.
* Source/WebCore/dom/Subscriber.cpp: Added.
(WebCore::Subscriber::create):
(WebCore::Subscriber::Subscriber):
(WebCore::Subscriber::next):
(WebCore::Subscriber::error):
(WebCore::Subscriber::complete):
(WebCore::Subscriber::addTeardown):
(WebCore::Subscriber::followSignal):
(WebCore::Subscriber::close):
(WebCore::Subscriber::isInactiveDocument const):
(WebCore::Subscriber::reportErrorObject):
* Source/WebCore/dom/Subscriber.h: Added.
* Source/WebCore/dom/Subscriber.idl: Added.
* Source/WebCore/dom/SubscriberCallback.h: Added.
* Source/WebCore/dom/SubscriberCallback.idl: Added.
* Source/WebCore/dom/SubscriptionObserver.h: Added.
* Source/WebCore/dom/SubscriptionObserver.idl: Added.
* Source/WebCore/dom/SubscriptionObserverCallback.h: Added.
* Source/WebCore/dom/SubscriptionObserverCallback.idl: Added.

Canonical link: <a href="https://commits.webkit.org/281040@main">https://commits.webkit.org/281040@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4854990071d30b86cb521c50fe7891dd581d138b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58466 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37793 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10950 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62092 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8910 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60595 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9107 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47343 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6354 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60497 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35389 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50546 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28188 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7848 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7914 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/51557 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54094 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8119 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63795 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57708 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8141 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54665 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2387 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50572 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54732 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12908 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2008 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79469 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33622 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13226 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34708 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35792 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34453 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

keithamus commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.any-expected.txt:
```diff
@@ -1,38 +1,50 @@
+PASS Subscription does not emit values after completion
+PASS Subscription does not emit values after error
+PASS Completing or nexting a subscriber after an error does nothing
+FAIL Errors pushed to the subscriber that are not handled by the subscription are reported to the global assert_equals: Error message matches expected "Uncaught Error: custom error" but got "Error: custom error"
```
I think this failure is something that needs to be addressed in WPTs. WebKit doesn't prepend `Uncaught` in re-raised exceptions, generally. The WPTs assume it will (largely because they were written alongside the Chrome implementation). I believe this is also the case in Firefox.

keithamus commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.any-expected.txt:
```diff
@@ -1,38 +1,50 @@
+PASS Errors pushed by initializer function after subscriber is closed by error are reported
+PASS Subscriber#complete() cannot re-entrantly invoke itself
+PASS Subscriber#error() cannot re-entrantly invoke itself
+FAIL Unsubscription lifecycle assert_array_equals: expected property 1 to be "outer abort handler" but got "teardown 2" (expected array ["subscribe() callback", "outer abort handler", "teardown 2", "teardown 1", "inner abort handler", "abort() returned"] got ["subscribe() callback", "teardown 2", "teardown 1", "inner abort handler", "outer abort handler", "abort() returned"])
```
This one needs to be investigated further. If possible, I would rather merge we what currently have and follow up to resolve the ordering issue here in a follow up PR.

keithamus commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.window-expected.txt:
```diff
@@ -1,6 +1,6 @@
+PASS No observer handlers can be invoked in detached document
+PASS Subscriber.error() does not "report the exception" even when an `error()` handler is not present, when it is invoked in a detached document
+PASS Cannot subscribe to an Observable in a detached document
 FAIL Observable from EventTarget does not get notified for events in detached documents promise_test: Unhandled rejection with value: object "TypeError: event_target.on is not a function. (In 'event_target.on('customevent')', 'event_target.on' is undefined)"
```
To be implemented in a follow up PR, to keep review-size down.

keithamus commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-takeUntil.any-expected.txt:
```diff
@@ -1,13 +1,13 @@
-FAIL takeUntil: source is uninterrupted when notifier completes, even synchronously promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
-FAIL takeUntil() mirrors the source Observable until its first next() value promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
-FAIL takeUntil: notifier calls `Subscriber#error()` twice; second goes to global error handler promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
+FAIL takeUntil subscribes to source Observable and mirrors it uninterrupted promise_test: Unhandled rejection with value: object "TypeError: source.takeUntil is not a function. (In 'source.takeUntil(new Observable(() => {}))', 'source.takeUntil' is undefined)"
```
To be implemented in a follow up PR, to keep review-size down.

keithamus commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-takeUntil.any.worker-expected.txt:
```diff
@@ -1,13 +1,13 @@
-FAIL takeUntil: source is uninterrupted when notifier completes, even synchronously promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
-FAIL takeUntil() mirrors the source Observable until its first next() value promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
-FAIL takeUntil: notifier calls `Subscriber#error()` twice; second goes to global error handler promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
+FAIL takeUntil subscribes to source Observable and mirrors it uninterrupted promise_test: Unhandled rejection with value: object "TypeError: source.takeUntil is not a function. (In 'source.takeUntil(new Observable(() => {}))', 'source.takeUntil' is undefined)"
```
To be implemented in a follow up PR, to keep review-size down.

keithamus commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any-expected.txt:
```diff
@@ -1,8 +1,8 @@
-FAIL toArray(): Subscribing with an aborted signal returns an immediately rejected promise promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
-FAIL toArray(): Aborting the passed-in signal rejects the returned promise promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
-FAIL Operator Promise abort ordering promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
+FAIL toArray(): basic next/complete promise_test: Unhandled rejection with value: object "TypeError: observable.toArray is not a function. (In 'observable.toArray()', 'observable.toArray' is undefined)"
```
To be implemented in a follow up PR, to keep review-size down.

keithamus commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-takeUntil.window-expected.txt:
```diff
@@ -1,4 +1,4 @@
 
-FAIL takeUntil(): notifier Observable detaches document before source Observable would be subscribed to promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
-FAIL takeUntil(): Source and notifier internal observers do not crash in a detached document promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
+FAIL takeUntil(): notifier Observable detaches document before source Observable would be subscribed to promise_test: Unhandled rejection with value: object "TypeError: source.takeUntil is not a function. (In 'source.takeUntil(notifier)', 'source.takeUntil' is undefined)"
```
To be implemented in a follow up PR, to keep review-size down.

keithamus commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any.worker-expected.txt:
```diff
@@ -1,8 +1,8 @@
-FAIL toArray(): Subscribing with an aborted signal returns an immediately rejected promise promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
-FAIL toArray(): Aborting the passed-in signal rejects the returned promise promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
-FAIL Operator Promise abort ordering promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: Observable"
+FAIL toArray(): basic next/complete promise_test: Unhandled rejection with value: object "TypeError: observable.toArray is not a function. (In 'observable.toArray()', 'observable.toArray' is undefined)"
```
To be implemented in a follow up PR, to keep review-size down.

rreno commented on Source/WebCore/dom/SubscriptionObserverCallback.h:
```diff
@@ -0,0 +1,43 @@
+public:
+    using ActiveDOMCallback::ActiveDOMCallback;
+
+    virtual bool hasCallback() const = 0;
```
This can be private as it shouldn't be called polymorphically.

rreno commented on Source/WebCore/dom/SubscriberCallback.h:
```diff
@@ -0,0 +1,45 @@
+public:
+    using ActiveDOMCallback::ActiveDOMCallback;
+
+    virtual bool hasCallback() const = 0;
```
Same, this can be private as it shouldn't be called polymorphically.

rreno commented on Source/WebCore/dom/Subscriber.h:
```diff
@@ -0,0 +1,77 @@
+    bool m_active = true;
+    Ref<AbortController> m_abortController;
+    RefPtr<ScriptExecutionContext> m_context;
+    RefPtr<JSSubscriptionObserverCallback> m_next;
```
You'll need some mechanism to keep these callback objects alive. As it stands they will likely be garbage collected even with these references.

I would suggest using `visitAdditionalChildren` for this interface as probably the easiest method to manage the lifetime.

rreno replied:
You also should be able to hold pointers to the base class rather than the derived `JS` variant.

keithamus commented on Source/WebCore/dom/Subscriber.h:
```diff
@@ -0,0 +1,89 @@
+        visitor.append(thisObject->m_complete);
+        for (auto iter = thisObject->m_teardowns.rbegin(); iter != thisObject->m_teardowns.rend(); ++iter)
+            visitor.append((*iter).get());
+    }
```
@rreno Here's the `visitAdditionalChildren` implementation. Does this look right to you?

annevk commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -4750,6 +4750,20 @@ NotificationsEnabled:
+  status: testable
+  category: dom
+  humanReadableName: "Observable API"
+  humanReadableDescription: "Enable the Observable API"
```
Support for the*

annevk commented on Source/WebCore/bindings/scripts/test/TestCallbackOrDict.idl:
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (C) 2024 Keith Cirkel <webkit@keithcirkel.co.uk>. All rights reserved.
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
I don't think you should be adding Apple Inc. unless this was taken from elsewhere? (In which case it should appear first.)

rreno reviewed (changes requested):
Before diving into the C++ code itself, I think we need to figure out what the lifetime of both the `Observable` and `Subscriber` objects should be. 

e.g. something like this will be immediately garbage collected (for some definition of 'immediate'):

```javascript
function makeObservableWithCallbackAndSubscribe(cb, subscribe)
{
    let observable = new Observable(cb);
    observable.subscribe(subscribe);
}

// ... later ...
makeObservableWithCallbackAndSubscribe(myCallback, mySubscribe);
setTimeout(() => continueProcessing(), 100);
```

In the above contrived example we want to subscribe to some observable events and wait some amount of time before continuing to do the work we're doing.

My question is: should the Observable stay alive in this situation? It currently has an outstanding subscriber that it should be feeding events to, even though there is no reference to it in JavaScript.

rreno commented on Source/WebCore/bindings/js/JSSubscriberCustom.cpp:
```diff
@@ -0,0 +1,46 @@
+    if (auto* complete = wrapped().completeCallbackConcurrently())
+        complete->visitJSFunction(visitor);
+    for (auto* teardown : wrapped().teardownCallbacksConcurrently())
+        teardown->visitJSFunction(visitor);
```
The returned type here is a vector of callbacks so they'll need to be iterated through keeping in mind that you'll need a lock to ensure thread safety.

rreno commented on Source/WebCore/dom/Subscriber.h:
```diff
@@ -0,0 +1,111 @@
+    std::vector<VoidCallback*> teardownCallbacksConcurrently()
+    {
+        // Copy into a vector to unpack Ref<>s
+        std::vector<VoidCallback*> callbacks;
```
Please use a `WTF::Vector`. Also this should be made thread-safe with a lock. Something like this would work if you turn `m_teardowns` into a Vector as well:

```C++
Vector<VoidCallback*> teardownCallbacksCopy() const
{
    Locker locker { m_teardownsLock };
    return m_teardowns.map([](auto& callback) { return callback.ptr(); });
}
```

rreno replied:
By making a copy thread-safely here you won't need a lock in the `JSSubscriberCustom` code above.

rreno commented on Source/WebCore/dom/Subscriber.h:
```diff
@@ -0,0 +1,111 @@
+    RefPtr<SubscriptionObserverCallback> m_next;
+    RefPtr<SubscriptionObserverCallback> m_error;
+    RefPtr<VoidCallback> m_complete;
+    ListHashSet<Ref<VoidCallback>> m_teardowns;
```
Why a `ListHashSet`? We don't ever look up a callback, only ever appending or iterating through the whole structure. I think a `Vector` would work fine for this.

rreno replied:
To make it thread safe you can add a `mutable Lock` and then annotate this with a `WTF_GUARDED_BY_LOCK(m_teardownsLock);` so thread safety analysis works.

rreno commented on Source/WebCore/dom/Subscriber.h:
```diff
@@ -0,0 +1,111 @@
+        m_next = nullptr;
+        m_error = nullptr;
+        m_complete = nullptr;
+        m_teardowns.clear();
```
We'd need every access to `m_teardowns` to be thread-safe.

rreno commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -0,0 +1,162 @@
+    if (m_active)
+        m_teardowns.add(callback);
+    else
+        callback.get().handleEvent();
```
Instead of using `get` whenever you're dereferencing a `Ref` or a `RefPtr` you can use `operator->`. So:

```C++
callback->handleEvent()
```

rreno commented on Source/WebCore/dom/Subscriber.idl:
```diff
@@ -0,0 +1,33 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+[Exposed=(Window,Worker), ActiveDOMObject, JSCustomMarkFunction, EnabledBySetting=ObservableEnabled]
```
For multiple attributes we usually format them one per line like this:


```
[
    Exposed=(Window,Worker),
    ActiveDOMObject,
    JSCustomMarkFunction,
    EnabledBySetting=ObservableEnabled
]
interface Subscriber {
```

rreno commented on Source/WebCore/dom/SubscriberCallback.idl:
```diff
@@ -0,0 +1,26 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+[RethrowException] callback SubscriberCallback = undefined (Subscriber subscriber);
```
Also we add a space between the attribute and `[]`.
```
[ RethrowException ] callback SubscriberCallback ...
```

rreno commented on Source/WebCore/dom/Observable.h:
```diff
@@ -0,0 +1,60 @@
+    WTF_MAKE_ISO_ALLOCATED(Observable);
+
+public:
+    using ObserverUnion = std::variant<RefPtr<JSSubscriptionObserverCallback>, SubscriptionObserver>;
```
I'd do a `RefPtr<SubscriptionObserverCallback>` rather than the `JS` version unless you've got a good reason for the latter?

keithamus replied:
Refused to compile and I'm not sure exactly how to solve it:

```
In file included from /app/webkit/WebKitBuild/GTK/Debug/JavaScriptCore/PrivateHeaders/JavaScriptCore/ThrowScope.h:28,
/app/webkit/WebKitBuild/GTK/Debug/WebCore/DerivedSources/JSObservable.cpp: In lambda function:
/app/webkit/WebKitBuild/GTK/Debug/WebCore/DerivedSources/JSObservable.cpp:224:206: error: cannot convert ‘optional<variant<WTF::RefPtr<WebCore::JSSubscriptionObserverCallback, WTF::RawPtrTraits<WebCore::JSSubscriptionObserverCallback>, WTF::DefaultRefDerefTraits<WebCore::JSSubscriptionObserverCallback> >, WebCore::SubscriptionObserver>>’ to ‘optional<variant<WTF::RefPtr<WebCore::SubscriptionObserverCallback, WTF::RawPtrTraits<WebCore::SubscriptionObserverCallback>, WTF::DefaultRefDerefTraits<WebCore::SubscriptionObserverCallback> >, WebCore::SubscriptionObserver>>’
  224 |     RELEASE_AND_RETURN(throwScope, JSValue::encode(toJS<IDLUndefined>(*lexicalGlobalObject, throwScope, [&]() -> decltype(auto) { return impl.subscribe(*context, observerConversionResult.releaseReturnValue(), optionsConversionResult.releaseReturnValue()); })));
/app/webkit/WebKitBuild/GTK/Debug/JavaScriptCore/PrivateHeaders/JavaScriptCore/ExceptionScope.h:126:16: note: in definition of macro ‘RELEASE_AND_RETURN’
  126 |         return expression__; \
In file included from /app/webkit/WebKitBuild/GTK/Debug/WebCore/DerivedSources/JSObservable.h:24,
/app/webkit/Source/WebCore/dom/Observable.h:52:45: note:   initializing argument 2 of ‘void WebCore::Observable::subscribe(WebCore::ScriptExecutionContext&, std::optional<std::variant<WTF::RefPtr<WebCore::SubscriptionObserverCallback, WTF::RawPtrTraits<WebCore::SubscriptionObserverCallback>, WTF::DefaultRefDerefTraits<WebCore::SubscriptionObserverCallback> >, WebCore::SubscriptionObserver> >, WebCore::SubscribeOptions)’
   52 |     void subscribe(ScriptExecutionContext&, std::optional<ObserverUnion>, SubscribeOptions);
In file included from /app/webkit/Source/WebCore/bindings/js/JSDOMConvertDictionary.h:29,
/app/webkit/Source/WebCore/bindings/js/JSDOMConvertBase.h: In instantiation of ‘JSC::JSValue WebCore::toJS(JSC::JSGlobalObject&, U&&) [with IDL = IDLUndefined; U = jsObservablePrototypeFunction_subscribeBody(JSC::JSGlobalObject*, JSC::CallFrame*, IDLOperation<JSObservable>::ClassParameter)::<lambda()>]’:
/app/webkit/Source/WebCore/bindings/js/JSDOMConvertBase.h:195:29:   required from ‘JSC::JSValue WebCore::toJS(JSC::JSGlobalObject&, JSC::ThrowScope&, U&&) [with IDL = IDLUndefined; U = jsObservablePrototypeFunction_subscribeBody(JSC::JSGlobalObject*, JSC::CallFrame*, IDLOperation<JSObservable>::ClassParameter)::<lambda()>]’
/app/webkit/WebKitBuild/GTK/Debug/WebCore/DerivedSources/JSObservable.cpp:224:5:   required from here
/app/webkit/Source/WebCore/bindings/js/JSDOMConvertBase.h:156:47: error: incomplete type ‘WebCore::JSConverter<WebCore::IDLUndefined>’ used in nested name specifier
  156 |     return JSConverterOverloader<IDL>::convert(lexicalGlobalObject, std::forward<U>(value));
[15/74] Building CXX object Source/WebCore/CMakeFiles/WebCore.dir/__/__/WebCore/DerivedSources/unified-sources/UnifiedSource-3a52ce78-38.cpp.o
ninja: build stopped: subcommand failed.
```

rreno replied:
This might be a bug in the code generator and/or support code for the generator or its a limitation of variant. In the generated code for `Observable.subscribe` it seems like it cant cast from `variant<RefPtr<[DerivedClass]>, Other>` to `variant<RefPtr<[BaseClass]>, Other>` which seems wrong. It could be due to the template monstrosity that is created. You should always be able to cast from pointer to derived to pointer to base.

rreno replied:
So in the mean time this is fine but we should figure out what's going on here.

rreno replied:
Looks like it might just be a limitation of variant since this doesn't compile:

```C++
#include <variant>

struct A { };
struct B : public A { };

using MyUnion = std::variant<A*, int>;

std::variant<B*, int> makeObject()
{
    return new B();
}

int main()
{
    MyUnion mu = makeObject();
    return 0;
}
```

I'll look into fixing the code generator in the future. I filed this: https://bugs.webkit.org/show_bug.cgi?id=276665

tuankiet65 replied:
I'm hitting this bug while implementing [this](https://drafts.csswg.org/css-view-transitions-2/#extend-document-types) (extension to `document.startViewTransition()`). If nobody is looking into that bugzilla, I could take over it.

rreno commented on Source/WebCore/dom/SubscriptionObserverCallback.idl:
```diff
@@ -0,0 +1,26 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+callback SubscriptionObserverCallback = undefined (any value);
```
I know this isn't your doing but rather the wording of the spec but this name is so confusing to me. Why not ObservableCallback or ObservableSubscriptionCallback? By prefacing with Subscription it clashes with the Subscriber interface and callbacks and it's been tough to keep it straight in my head.

`ObservableSubcscriptionCallback` is my favorite. Maybe worth filing a spec issue to bikeshed the name a bit.

keithamus replied:
https://github.com/WICG/observable/pull/157

rreno commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -0,0 +1,100 @@
+Ref<Subscriber> Observable::makeSubscriber(ScriptExecutionContext& context, std::optional<ObserverUnion> observer)
+{
+    if (observer.has_value()) {
+        return std::visit(
```
This is just `WTF::switchOn`. You'd write it like this:

```C++
return switchOn(observer.value(),
    [&](RefPtr<JSSubscriptionObserverCallback>& next) {
        // ...
    },
    // ...
```

rreno commented on Source/WebCore/dom/Observable.idl:
```diff
@@ -0,0 +1,32 @@
+// https://github.com/WICG/observable
+typedef (SubscriptionObserverCallback or SubscriptionObserver) ObserverUnion;
+
+[Exposed=(Window,Worker), EnabledBySetting=ObservableEnabled]
```
Multiple attributes, please format on individual lines.

rreno commented on Source/WebCore/dom/Observable.idl:
```diff
@@ -0,0 +1,32 @@
+[Exposed=(Window,Worker), EnabledBySetting=ObservableEnabled]
+interface Observable {
+  [RaisesException] constructor(SubscriberCallback callback);
+  [CallWith=CurrentScriptExecutionContext, RaisesException] undefined subscribe(optional ObserverUnion observer = {}, optional SubscribeOptions options = {});
```
These don't need to be formatted on multiple lines though, just the top level.

rreno commented on Source/WebCore/dom/Observable.h:
```diff
@@ -0,0 +1,61 @@
+class ScriptExecutionContext;
+class JSSubscriptionObserverCallback;
+class SubscriptionObserverCallback;
+class SubscriptionObserver;
```
This and the next forward declaration need to be `struct`, since they are ultimately defined as `struct`s elsewhere.

rreno commented on Source/WebCore/dom/Observable.h:
```diff
@@ -0,0 +1,61 @@
+
+namespace WebCore {
+
+class Document;
```
I think this forward declaration is unnecessary.

rreno commented on Source/WebCore/dom/Subscriber.cpp:
```diff
@@ -0,0 +1,171 @@
+    if (isInactiveDocument())
+        return;
+
+    if (m_active)
```
Missing a lock in this scope.

rreno reviewed (approved):
This is looking good! Once the bots are green and the last of the review comments are resolved I think it's ready to land in a testable state. Thanks!

keithamus committed:
Implement Observable class and subscribe method.
https://bugs.webkit.org/show_bug.cgi?id=276572

Reviewed by Ryan Reno.

This implements the base `Observable` class, as well as the `subscribe`
method. In addition this adds the `Subscriber` internal class allows
Observables to be subscribed to, and some additional machinery such as
`SubscribeOptions`, `SubscriptionObserver` and so on.

This also includes a fix to IDL generation to correctly generate callback
function + dictionary unions (https://github.com/whatwg/webidl/pull/1353).

This does not implement any of the prototype methods on Observable,
(other than subscribe) such as `takeUntil`, `filter`, `finally` and so
on.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-constructor.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-takeUntil.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-takeUntil.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-takeUntil.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-toArray.any.worker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/JSDOMConvertUnion.h:
* Source/WebCore/bindings/js/JSSubscriberCustom.cpp: Added.
(WebCore::JSSubscriber::visitAdditionalChildren):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(AreTypesDistinguishableForOverloadResolution):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackWithFunctionOrDict.cpp: Added.
(WebCore::JSTestCallbackWithFunctionOrDict::JSTestCallbackWithFunctionOrDict):
(WebCore::JSTestCallbackWithFunctionOrDict::~JSTestCallbackWithFunctionOrDict):
(WebCore::JSTestCallbackWithFunctionOrDict::handleEvent):
(WebCore::JSTestCallbackWithFunctionOrDict::visitJSFunction):
(WebCore::toJS):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackWithFunctionOrDict.h: Added.
(WebCore::toJS):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestCallbackOrDict.idl: Added.
* Source/WebCore/dom/Observable.cpp: Added.
(WebCore::Observable::create):
(WebCore::Observable::subscribe):
(WebCore::Observable::makeSubscriber):
(WebCore::Observable::Observable):
* Source/WebCore/dom/Observable.h: Added.
* Source/WebCore/dom/Observable.idl: Added.
* Source/WebCore/dom/SubscribeOptions.h: Added.
* Source/WebCore/dom/SubscribeOptions.idl: Added.
* Source/WebCore/dom/Subscriber.cpp: Added.
(WebCore::Subscriber::create):
(WebCore::Subscriber::Subscriber):
(WebCore::Subscriber::next):
(WebCore::Subscriber::error):
(WebCore::Subscriber::complete):
(WebCore::Subscriber::addTeardown):
(WebCore::Subscriber::followSignal):
(WebCore::Subscriber::close):
(WebCore::Subscriber::isInactiveDocument const):
(WebCore::Subscriber::reportErrorObject):
* Source/WebCore/dom/Subscriber.h: Added.
* Source/WebCore/dom/Subscriber.idl: Added.
* Source/WebCore/dom/SubscriberCallback.h: Added.
* Source/WebCore/dom/SubscriberCallback.idl: Added.
* Source/WebCore/dom/SubscriptionObserver.h: Added.
* Source/WebCore/dom/SubscriptionObserver.idl: Added.
* Source/WebCore/dom/SubscriptionObserverCallback.h: Added.
* Source/WebCore/dom/SubscriptionObserverCallback.idl: Added.

Canonical link: https://commits.webkit.org/281040@main
